### ISSUE TITLE
Fetch calls have no timeout and can stall indefinitely

### ISSUE THREAD

**Issue by baileyherbert** [TTTaevas/osu-api-v2-js#32]
Hi, thanks for your work on this awesome package and congrats on the v1 release! :grin:

The current implementation can lead to indefinite stalls for requests, because `fetch()` does not have a timeout by default, and the options configured in this package do not implement one. This will rarely pose problems, but after using this heavily on some test projects over the last few days, I've now encountered those stalls twice. 😓 

The easiest approach would be to add a basic timeout signal to the fetch options for requests and authentication, e.g.:

```ts
fetch(url, {
    ...,
    signal: AbortSignal.timeout(30000)
})
```

_It should be noted that this `AbortSignal.timeout()` method was introduced in Node [v17.3](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/timeout_static#browser_compatibility). For older versions, the alternative would be to create an `AbortController` and abort it manually with a timeout._ 

It would be ideal for the timeout to be configurable in the client itself, and for `API.createAsync()` to accept it as an option. Personally, I think it would be even better if each API method accepted an optional signal so developers could cancel the requests for other reasons too, but that seems like it would be messy to implement here as some methods already have many parameters.

I would be happy to work on a pull request for this, but I would be keen to hear which direction you want to go.

**Comment by TTTaevas** [comment]
Hello, thank you for using this package and reporting this issue! :D

Sorry you've had to experience those stalls, I will soon be adding a configurable timeout, probably using an `AbortController` to ensure compatibility (even if node versions prior to 18 are no longer supported)

I understand why you'd want any request to be cancelled at any time, and as you said, having every methods optionally accept a signal for that sole purpose would quickly make things a mess
However, there could be a way for users to call signals, by making the `AbortController` accessible through the `api` object

For example, right before calling `fetch`, the `api` may create and add an `AbortController` to a new property called `controllers` which would be an Array of `AbortController`s, and remove it from the array once the request is finished (completed or cancelled)
This approach would be problematic if you're looking to abort a specific request, though

A similar approach would be instead to have a property that lists the details of all the current requests alongside an `AbortController`
The `api` would have an array called `current_requests`, composed of objects like

```ts
{
 controller: AbortController,
 method: "get" | "post" | "put" | "delete",
 endpoint: string,
 parameters: {[k: string]: any}
}
```

Before I do anything like that, let me know if you think this could work well for the purpose of cancelling requests!

**Comment by baileyherbert** [comment]
Thanks @TTTaevas! 

For what it's worth, I'm not personally a fan of either approach. It's a bit too unpredictable to obtain the correct controller. I also anticipate that in most cases, developers will want to group multiple requests (if not all) under the same abort signal, and sometimes, multiple groups of requests with different signals. Therefore, it would probably be best if they provided their own.

I've given the problem of implementation considerable thought and have some ideas for it. However, the fine details might get complicated. IMO, regardless of the ability to provide custom signals, there should still at least be a configurable timeout that applies to all requests, and it should still apply even when users (eventually) provide their own abort signals, so I propose we start simple with just the timeout part first. Of course, it's ultimately up to you. :wink:

Proposed example usage:

```ts
// Initial auth uses a default timeout (e.g. 30000)
const client = await osu.API.createAsync({
    id: 1234,
    secret: '**********'
});

// Can then be overridden
client.timeout = 15000;
```

Here's an example fetch wrapper that should work wonderfully for this purpose, though I have not tested it, assuming a `timeout` field is added to the class. Later, custom signals can be implemented here too, but some minor work would be required to merge the two signals together as we can only configure one.

```ts
private async fetchWithTimeout(url: RequestInfo, init: RequestInit) {
	if (this.timeout <= 0) {
		return fetch(url, init);
	}

	const abort_controller = new AbortController();
	const timeout_object = setTimeout(() => {
		abort_controller.abort();
	}, this.timeout);

	try {
		return await fetch(url, {
			...init,
			signal: abort_controller.signal
		});
	}
	finally {
		clearTimeout(timeout_object);
	}
}
```

Timeouts will throw an `AbortError` back up to the user, which they will need to handle on their own.

There should be a reasonable default timeout (perhaps `30000`) that the authentication flow behind `API.createAsync()` will use, and setting the timeout to `0` (or less, as a protective measure) should disable it, which will be useful later when custom abort signals can be provided.

What do you think? 🙂

**Comment by TTTaevas** [comment]
Yes, I agree with what you've said regarding how abort signals should be provided! I think discussion regarding that should be in another issue, so we keep this one relevant only to the matter of timeouts :p

So regarding timeout stuff, I also thought of it on my side, reaching very similar conclusions!

Much like an `api`'s `server`, `timeout` will be configurable both at the time of creation of the `api` object and when overridden at any given time by the user
It will be in seconds with a default of `20` and may be set to `0` or less to be deactivated altogether

```ts
const controller = new AbortController()
const timer = this.timeout > 0 ? setTimeout(() => {
	to_retry = true // specifically retry the request if the reason for abortion is the timeout
	controller.abort()
}, this.timeout * 1000) : false

const response = await fetch(url, {
	/** <irrelevant fetch code erased> **/
	signal: controller.signal
})
/** <irrelevant error-handling code erased> **/
		
if (timer) {
	clearTimeout(timer)
}
```

Upon timing out, it will throw an [`APIError`](https://osu-v2.taevas.xyz/classes/APIError.html) with a `message` `"AbortError (aborted)"`, unless it's during `API.createAsync()` in which case it'd simply be an `AbortError` (i think, not tested)

I must admit not being sure about setting `to_retry` to `true` which makes the `api` try this specific request again instead of throwing an error, do you think it would be welcome or would it instead get in the way? :o 

I'm otherwise feeling pretty confident in that code! Feel free to let me know what you think
And thank you for helping me out!!

**Comment by baileyherbert** [comment]
Looks great.

Honestly, I think the retry logic should be disabled by default for timeouts, but potentially made optional. The reason is if these requests are being sent on behalf of an end user, there might be a genuine timeframe that the request must be fulfilled by, and the retry feature will break that.

I would also caution you about when `fetch` throws an error, at least with the snippet you provided, the timeout would not be cleared. This would cause it to "leak" and could prevent the process from shutting down gracefully until it clears naturally. That's easy to fix by calling `unref()` on the `timer` object, but the retry logic will further complicate this.

In my situation, I call these API methods from within a loop and will already retry on my own when something fails.

Finally, for future proofing, it's often essential to easily distinguish between an abort error and other general errors. At the very least I would strongly recommend a helper method on the `APIError` class to distinguish it, maybe `isAbortError(): boolean`?

Anyway, yeah, it looks great overall! :+1:

**Comment by TTTaevas** [comment]
Yep, I've added something to `APIError` to make it easier to distinguish certain errors from others, good call there :p
With the release of v1.0.1, I think it's safe to close this issue!!
